src/core/binding.c

```diff
@@ -1380,6 +1380,51 @@ QuicBindingCreateConnection(
     return Connection;
 }
 
+_IRQL_requires_max_(DISPATCH_LEVEL)
+BOOLEAN
+QuicBindingDropBlockedSourcePorts(
+    _In_ QUIC_BINDING* Binding,
+    _In_ const CXPLAT_RECV_DATA* const Datagram
+    )
+{
+    const uint16_t SourcePort = QuicAddrGetPort(&Datagram->Route->RemoteAddress);
+
+    //
+    // These UDP source ports are recommended to be blocked by the QUIC WG. See
+    // draft-ietf-quic-applicability for more details.
```

Comment:
I looked at the latest version of draft-ietf-quic-applicability. Section 8.1 ("Source Port Selection") seems to be the relevant part. That section explicitly mentions that serverside blocking of these ports is problematic, and therefore recommends clients avoid the use of these ports for its source port.

But here you seem to be doing serverside port blocking?

Aside from that, isn't this type of thing a job for a firewall? IMO a filter like this does not belong in core QUIC.

Reply:
As I briefly mentioned in the description, I made this PR based on a Slack discussion with two of the largest deployers of QUIC worldwide: Google and Facebook. Both have taken this path of blocking things on the server side. The QUIC protocol has known issues with a couple of these other protocols that can cause infinite loops. The rest have been seen in actual DoS attacks in the wild.

> ... therefore recommends clients avoid the use of these ports for its source port.

We could also block on the client side, but that doesn't protect us when we don't own the client (an attacker isn't going to respect that). Also, it would make testing this harder (but that's not really the point).

> Aside from that, isn't this type of thing a job for a firewall? IMO a filter like this does not belong in core QUIC.

Perhaps a firewall could be responsible for blocking these ports, but until we have such protections in place (especially in Azure) QUIC is the only place we have. Additionally, since there are known protocol incompatibilities with some of these protocols, the QUIC protocol is a good place to handle these issues IMO.

Reply:
Have you brought this issue to Azure? I'd like to know if Google and Facebook used firewalls to block these source ports or if they put filters in their core QUIC implementations.

Reply:
> Have you brought this issue to Azure?

I have not brought this specific issue of these ports being blocked to Azure yet. But it will be discussed as a part of the larger DDoS protection effort in Azure.

> I'd like to know if Google and Facebook used firewalls to block these source ports or if they put filters in their core QUIC implementations.

They do block the ports directly in their source.

Reply:
I tweaked the comment. Let me know if that looks better.

Reply:
Yeah I think that's better